### ISSUE TITLE
White flash on initial open

### ISSUE THREAD

**Issue by mustaqimM** [doomemacs/doomemacs#2828]
**What did you expect to happen?**
No white flash when a dark theme is set

**What actually happened?**
A white flash appears after opening (I use emacs in daemon mode so I only see it once, otherwise it occurs on every open ). I don't think it used to do this.

**Additional details:**
Here's a [video](https://streamable.com/7pcddl) of it happening
My [config](https://github.com/mustaqimM/dotfiles/blob/master/.config/doom/config.el)

**Steps to reproduce:**
1. Run `/bin/emacs --daemon`
2. Open directly.

**System information:**
<details><pre>
<details>

```
emacs   version    26.3
        features   XPM JPEG TIFF GIF PNG RSVG SOUND DBUS GSETTINGS GLIB NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS LCMS2
        build      Aug 29, 2019
        buildopts  (--with-x-toolkit=gtk3 --with-xwidgets --prefix=/usr --sysconfdir=/etc --sbindir=/usr/bin --bindir=/usr/bin --mandir=/usr/share/man --infodir=/usr/share/info --localstatedir=/var --with-file-notification=inotify --with-modules --with-jpeg --with-tiff --with-gif --with-png --with-xpm --with-rsvg --without-imagemagick --with-xml2 --with-gnutls --with-sound --with-m17n-flt --host=x86_64-unknown-linux-gnu --build=x86_64-unknown-linux-gnu 'CFLAGS=-fno-PIE -mtune=generic -O2 -pipe   -g' 'CPPFLAGS=   ' 'LDFLAGS=-no-pie -Wl,--as-needed    ')
        windowsys  x
        daemonp    daemon
doom    version    2.0.9
        build      HEAD -> develop a381f5926 2020-04-03 02:12:50 -0400
        dir        ~/.config/doom/
system  type       gnu/linux
        config     x86_64-unknown-linux-gnu
        shell      /bin/sh
        uname      Linux 5.4.27_1 #1 SMP PREEMPT Sat Mar 21 09:34:54 UTC 2020 x86_64
        path       (~/.zinit/snippets/cht.sh ~/.zinit/plugins/zdharma---zsh-diff-so-fancy/bin ~/.zinit/plugins/junegunn---fzf-bin ~/.zinit/plugins/imsnif---bandwhich ~/.zinit/plugins/karan---joe ~/.zinit/plugins/dylanaraps---pfetch ~/.zinit/plugins/Peltoche---lsd/lsd ~/.zinit/plugins/pixelb---ps_mem ~/.zinit/plugins/direnv---direnv ~/.zinit/plugins/sei40kr---fast-alias-tips-bin ~/.zinit/polaris/bin ~/.cargo/bin ~/.local/bin/sml/bin ~/.yarn/bin ~/.emacs.d/bin /usr/local/bin /bin /usr/bin /usr/local/sbin /usr/sbin /sbin /opt/texlive/2019/bin/x86_64-linux /usr/libexec/emacs/26.3/x86_64-unknown-linux-gnu/)
config  envfile    envvar-file
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash ophints (popup +all +defaults) tabs treemacs vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold multiple-cursors rotate-text snippets :emacs dired electric ibuffer vc :checkers syntax :tools editorconfig (eval +overlay) (lookup +docsets) magit rgb :lang data emacs-lisp latex lua markdown (org +dragndrop +present) plantuml python ruby rust sh web :config (default +bindings +smartparens))
        packages   ((flycheck-inline) (sml-mode))
        unpin      (n/a)
        elpa       (n/a)
```
</details>
</pre></details>

**Comment by hlissner** [comment]
I cannot reproduce this. I suspect it's something to do with your particular build of Emacs or the way Emacs is launched. What Linux distro are you using?

> My config

Side note, [this usage of `after!` is incorrect and should throw an error](https://github.com/mustaqimM/dotfiles/blob/master/.config/doom/config.el#L61-L71). It should be:

```elisp
(after! centaur-tabs
  (defun centaur-tabs-hide-tab (x)
    ...))
```

**Comment by mustaqimM** [comment]
@hlissner I'm on Void Linux; on `emacs-gtk3-26.3_1`

Please close it if you wish, I usually run in it daemon mode (`/bin/emacs --daemon` on startup, the first time emacs is shown I see the flash but not on subsequent opens) so I'm not too concerned. There's just too many factors that could cause it.

(And thanks for the hint regarding my config!)

**Comment by mustaqimM** [comment]
@hlissner I tried on a re-install and it still occurs. I now launch with `/bin/emacs -bg '#17191a'` to get rid of that blinding white flash

**Comment by haluk** [comment]
I observe the same problem.

https://streamable.com/dyplrd

**Comment by mustaqimM** [comment]
Another way of getting rid of the flash when opening initially with `emacsclient` and you haven't started the daemon, is appending:
```lisp
(add-to-list 'default-frame-alist
             '(background-color . "#17191a"))
```
to your config.el.

**Comment by haluk** [comment]
> Another way of getting rid of the flash when opening initially with `emacsclient` and you haven't started the daemon, is appending:
> 
> ```lisp
> (add-to-list 'default-frame-alist
>              '(background-color . "#17191a"))
> ```
> 
> to your config.el.

It didn't work for me.

**Comment by hlissner** [comment]
Does this fix the issue for any of you?

```elisp
;; add to ~/.doom.d/config.el
(when (daemonp)
  (remove-hook 'after-make-frame-functions #'doom-init-theme-h)
  (add-hook 'server-after-make-frame-hook #'doom-init-theme-h))
```

**Comment by haluk** [comment]
> Does this fix the issue for any of you?
> 
> ```emacs-lisp
> ;; add to ~/.doom.d/config.el
> (when (daemonp)
>   (remove-hook 'after-make-frame-functions #'doom-init-theme-h)
>   (add-hook 'server-after-make-frame-hook #'doom-init-theme-h))
> ```

I added this to beginning of the file. It didn't fix. I also tried to add this to the end of file. It's the same.

**Comment by mustaqimM** [comment]
@hlissner It actually does the opposite:
![image](https://user-images.githubusercontent.com/24436622/81660052-d5d07580-9429-11ea-8ac0-9b3dbe8afba5.png)

It's now stuck like this and doesn't change to the set theme

**Comment by hlissner** [comment]
Is this still an issue?

**Comment by haluk** [comment]
Yes. I used to use Debian based system, I'm now using Arch based system and gccemacs. I still see a white screen before all packages are loaded.

**Comment by samarthkulshrestha** [comment]
> Yes. I used to use Debian based system, I'm now using Arch based system and gccemacs. I still see a white screen before all packages are loaded.

Yep. Same with me. At the bottom, it says something like 
```
Loading /etc/emacs/site-start.d/(other filenames)
```
But then it goes away and then Doom loads.

**Comment by shegeley** [comment]
I had this problem when I hadn't had the font specified in config file installed.

**Comment by micdiamond** [comment]
This blank white screen while waiting for doom to load up happened to me when I added this line to my config.el or initi.el
(find-file "~/org/todo.org")

I added this line to automatically open a todo org file every time I start Emacs.

**Comment by ItsNilDev** [comment]
try starting emacs with: `emacs -bg black/grey`

**Comment by hlissner** [comment]
This issue seems to occur because Emacs is packaged with site files that load at startup, which emit "Loading X" messages at startup (before `~/.emacs.d` is sourced). This forces Emacs to redraw early, at a time before any themes have loaded. This can also happen if you install packages through your OS package manager (packages like vterm or mu4e).

Not sure if this really Doom's jurisdiction to "fix", but you could try adding the following to `~/.emacs.d/early-init.el`:

```elisp
(advice-add 'load-file :override
            (lambda (file) (load (expand-file-name file) nil t t)))
```

**Comment by ItsNilDev** [comment]
Also, if you choose the theme from the emacs itself, it generates something like this in your init.el: 
```
'(custom-enabled-themes '(doom-gruvbox))
'(custom-safe-themes
 '("4d9473e372aa5916b8198f6b4ed596c66a7a3b30c64269f24850114514aa553b" default))
``` This makes emacs to be dark at first/opening

**Comment by rounin-rp** [comment]
This happened with me when I uncommented the `setq doom-font` part in my config.el file. The moment I commented it out again doom started working fine. Now the issue is I cannot change the font and its size in doom :(

**Comment by shegeley** [comment]
@rounin-rp have you read my previous comment? 
I wrote that I had this issue when the font specified in `doom-font` was not installed on my system. So you might check carefully if you specified the proper font in `doom-font` and it's being present on your system.

**Comment by rounin-rp** [comment]
> @rounin-rp have you read my previous comment? I wrote that I had this issue when the font specified in `doom-font` was not installed on my system. So you might check carefully if you specified the proper font in `doom-font` and it's being present on your system.

@shegeley  Yeah. Actually for me its not just font, I am not using doom-font. I was trying to use one of my system fonts inside doom and it failed. The same thing happened when I tried to add more changes. I tried to add `nyan-mode` on my doom emacs, and faced the same issue. No able to customize anything on `config.el` file :(

**Comment by port19x** [comment]
> **What did you expect to happen?** No white flash when a dark theme is set
> 
> **What actually happened?** A white flash appears after opening (I use emacs in daemon mode so I only see it once, otherwise it occurs on every open ). I don't think it used to do this.
> 
> **Additional details:** Here's a [video](https://streamable.com/7pcddl) of it happening My [config](https://github.com/mustaqimM/dotfiles/blob/master/.config/doom/config.el)
> 
> **Steps to reproduce:**
> 
> 1. Run `/bin/emacs --daemon`
> 2. Open directly.
> 
> **System information:**

Can confirm. Tho thats probably outside the control of doom.
I have a white rectangle where the window should be that then gets replaced with emacs once it loads. Probably related to xorg / emacs itself

**Comment by frankjonen** [comment]
I have it on macOS, too now. Just did a fresh install with emacs-plus 29.3.
Flashes maximum white for over 1 second.

In my `early-init.el` I have this

```elisp
(add-to-list 'default-frame-alist '(fullscreen . maximized))
(add-to-list 'default-frame-alist '(undecorated . t))
```

Which means I'm really getting the hammer to the face now.

It wouldn't surprise me if it's a font loading issue (I'm using JetBrains Mono which isn't that big though). When I load a big unicode font for org-mode, org takes forever to start.

IF it really is a font issue (27 didn't have that startup issue), maybe defer font loading to right before the main-screen renders?

**Comment by cyscarinci** [comment]
I have the same issue with a free install of doom emacs.  Is it possible to change the theme on early-init.el on .config/emacs to a dark background? Does doom upgrade change this file, or is it persistent?

**Comment by cyscarinci** [comment]
Just in case this helps anyone, you can fix the white flash by loading a dark theme in early-init. 

For this, add, for example,
(load-theme 'modus-vivendi t)
to your ~/.config/emacs/early-init.el 

doom upgrade, however, does not like this. So you can make a shell function like this

doom_upgrade() {
    doom upgrade --force
    sed -i '$i\\n\(load-theme '\''modus-vivendi t\)\n' ~/.config/emacs/early-init.el
}

I imagine one must be careful if other changes to files in ~/.config/emacs are made. doom upgrade --force might overwrite other things.

**Comment by hlissner** [comment]
I am closing this issue because it is unactionable, can't be reproduced consistently, and all proposed solutions I've seen work for some and not others. The one instance where I could reproduce it (due to site-lisp files emitting output in the Echo Area during startup, triggering a premature redraw) [was addressed](https://github.com/doomemacs/doomemacs/blob/master/lisp/doom.el#L871-L878), but the rest of the reports seem inconsistent across different versions and builds of Emacs and operating systems. Plus, I've also seen this reported before in vanilla Emacs or users on other starter kits (like Spacemacs), which leads me to suspect this is an external issue, either with your build of Emacs, your local configuration, or even your window/desktop environment.

I'll reopen this if I find any reason to suspect Doom is the cause.

**Comment by edmondburnett** [comment]
This started happening to me randomly after I installed mu4e and started configuring it.

Removing `(require 'mu4e)` from my `config.el` fixes it.
